flink-optimizer/src/main/java/org/apache/flink/optimizer/traversals/RangePartitionRewriter.java

```diff
@@ -0,0 +1,188 @@
+		sicPlanNode.setParallelism(1);
+		sicChannel.setTarget(sicPlanNode);
+		sipPlanNode.addOutgoingChannel(sicChannel);
+		this.plan.getAllNodes().add(sicPlanNode);
+
+		// 3. Use sampled data to build range boundaries.
+		RangeBoundaryBuilder rangeBoundaryBuilder = new RangeBoundaryBuilder(comparator, targetParallelism);
+		TypeInformation<CommonRangeBoundaries> rbTypeInformation = TypeExtractor.getForClass(CommonRangeBoundaries.class);
+		UnaryOperatorInformation rbOperatorInformation = new UnaryOperatorInformation(sourceOutputType, rbTypeInformation);
+		MapPartitionOperatorBase rbOperatorBase = new MapPartitionOperatorBase(rangeBoundaryBuilder, rbOperatorInformation, "RangeBoundaryBuilder");
+		MapPartitionNode rbNode= new MapPartitionNode(rbOperatorBase);
+		Channel rbChannel = new Channel(sicPlanNode, TempMode.NONE);
+		rbChannel.setShipStrategy(ShipStrategyType.FORWARD, channel.getDataExchangeMode());
+		SingleInputPlanNode rbPlanNode = new SingleInputPlanNode(rbNode, "RangeBoundary PlanNode", rbChannel, DriverStrategy.MAP_PARTITION);
+		rbPlanNode.setParallelism(1);
+		rbChannel.setTarget(rbPlanNode);
+		sicPlanNode.addOutgoingChannel(rbChannel);
+		this.plan.getAllNodes().add(rbPlanNode);
+
+		// 4. Take range boundaries as broadcast input and take the tuple of partition id and record as output.
+		AssignRangeIndex assignRangeIndex = new AssignRangeIndex(comparator);
+		TypeInformation<Tuple2> ariOutputTypeInformation = new TupleTypeInfo<>(BasicTypeInfo.INT_TYPE_INFO, sourceOutputType);
+		UnaryOperatorInformation ariOperatorInformation = new UnaryOperatorInformation(sourceOutputType, ariOutputTypeInformation);
+		MapPartitionOperatorBase ariOperatorBase = new MapPartitionOperatorBase(assignRangeIndex, ariOperatorInformation, "Assign Range Index");
+		MapPartitionNode ariNode= new MapPartitionNode(ariOperatorBase);
+		Channel ariChannel = new Channel(sourceNode, TempMode.NONE);
+		ariChannel.setShipStrategy(ShipStrategyType.FORWARD, channel.getDataExchangeMode());
+		SingleInputPlanNode ariPlanNode = new SingleInputPlanNode(ariNode, "AssignRangeIndex PlanNode", ariChannel, DriverStrategy.MAP_PARTITION);
+		ariPlanNode.setParallelism(sourceParallelism);
+		ariChannel.setTarget(ariPlanNode);
+		appendOutgoingChannels.add(ariChannel);
+		this.plan.getAllNodes().add(ariPlanNode);
+
+		channel.setSource(ariPlanNode);
+		ariPlanNode.addOutgoingChannel(channel);
+
+		NamedChannel broadcastChannel = new NamedChannel("RangeBoundaries", rbPlanNode);
+		broadcastChannel.setShipStrategy(ShipStrategyType.BROADCAST, channel.getDataExchangeMode());
+		broadcastChannel.setTarget(ariPlanNode);
+		List<NamedChannel> broadcastChannels = new ArrayList<>(1);
+		broadcastChannels.add(broadcastChannel);
+		ariPlanNode.setBroadcastInputs(broadcastChannels);
+
+		// 5. Remove the partition id.
+		PartitionIDRemoveWrapper partitionIDRemoveWrapper = new PartitionIDRemoveWrapper();
```

Comment:
The `PartitionIDRemoveWrapper` must be injected before the operator that should receive the partitioned input because that operator does not expect the `Tuple2`-wrapped input and will fail otherwise. 
So we need to append the `PartitionIDRemoveWrapper` to the `PartitionIDAssigner` with a Range-partition channel. The idAssigner has the same DOP as the source and the idRemove has the same DOP as the target node. Then the target node is appended to the idRemove node with a forward channel. The range partitioning will happen between the idAssigner and idRemover but will be preserved due to the forward channel and the equal DOP of idRemover and target node.